erigon-lib/state/domain.go

```diff
@@ -814,47 +800,41 @@ func (d *Domain) collate(ctx context.Context, step, txFrom, txTo uint64, roTx kv
 	}
 	comp := NewArchiveWriter(coll.valuesComp, d.compression)
 
-	keysCursor, err := roTx.CursorDupSort(d.keysTable)
-	if err != nil {
-		return Collation{}, fmt.Errorf("create %s keys cursor: %w", d.filenameBase, err)
-	}
-	defer keysCursor.Close()
-
-	var (
-		stepBytes = make([]byte, 8)
-		keySuffix = make([]byte, 256+8)
-		v         []byte
-
-		valsDup kv.CursorDupSort
-	)
+	stepBytes := make([]byte, 8)
 	binary.BigEndian.PutUint64(stepBytes, ^step)
-	valsDup, err = roTx.CursorDupSort(d.valsTable)
+	valsCursor, err := roTx.Cursor(d.valsTable)
 	if err != nil {
 		return Collation{}, fmt.Errorf("create %s values cursorDupsort: %w", d.filenameBase, err)
 	}
-	defer valsDup.Close()
+	defer valsCursor.Close()
 
-	for k, stepInDB, err := keysCursor.First(); k != nil; k, stepInDB, err = keysCursor.Next() {
+	kvs := []struct {
+		k, v []byte
+	}{}
+
+	for k, v, err := valsCursor.First(); k != nil; k, v, err = valsCursor.Next() {
 		if err != nil {
 			return coll, err
 		}
+		stepInDB := k[len(k)-8:]
 		if !bytes.Equal(stepBytes, stepInDB) { // [txFrom; txTo)
 			continue
 		}
 
-		copy(keySuffix, k)
-		copy(keySuffix[len(k):], stepInDB)
-
-		v, err = roTx.GetOne(d.valsTable, keySuffix[:len(k)+8])
-		if err != nil {
-			return coll, fmt.Errorf("find last %s value for aggregation step k=[%x]: %w", d.filenameBase, k, err)
-		}
+		kvs = append(kvs, struct {
+			k, v []byte
+		}{common.Copy(k[:len(k)-8]), common.Copy(v)})
+	}
 
-		if err = comp.AddWord(k); err != nil {
-			return coll, fmt.Errorf("add %s values key [%x]: %w", d.filenameBase, k, err)
+	sort.Slice(kvs, func(i, j int) bool {
```

Comment:
keys are sorted in db. are you sure need additional sort?

Reply:
also write to files did not required extra linear amount of ram

Reply:
ah i see. His intetion is to remove table we got sorted and use unsorted valsTable and sort it (with ETL i think)